elliptic_interface.cc

```diff
@@ -762,44 +761,76 @@ unsigned int EllipticInterfaceDLM<dim>::solve() {
   // mass matrix.
   auto invM = null_operator(M);
   auto invW = null_operator(M);
-  // SparseDirectUMFPACK M_inv_umfpack;
   SolverControl solver_control(100, 1e-12, false, false);
   SolverCG<TrilinosWrappers::MPI::Vector> cg_solver(solver_control);
-  // Using inverse of diagonal mass matrix (squared)
-  // TODO: check
+
   TrilinosWrappers::MPI::Vector inverse_diag_mass_squared(locally_owned_dofs_fg,
                                                           comm);
-  // for (unsigned int i = 0; i < dof_handler_fg.n_dofs(); ++i)
-  for (const auto i : locally_owned_dofs_fg)
+
+  for (const types::global_dof_index i : locally_owned_dofs_fg)
     inverse_diag_mass_squared(i) =
         1. / (mass_matrix_fg.diag_element(i) * mass_matrix_fg.diag_element(i));
   inverse_diag_mass_squared.compress(VectorOperation::insert);
 
-  // TODO: parallelize
-  DiagonalMatrix<Vector<double>> diag_inverse;
-  if (parameters.use_diagonal_inverse == true) {
-    // TODO: adapt
-    // diag_inverse.reinit(inverse_diag_mass_squared);
-    // invW = linear_operator(diag_inverse);
-    // invM
-  } else {
-    // Use direct inversion
+  TrilinosWrappers::SparseMatrix diag_inverse;
+  TrilinosWrappers::SolverDirect solver_direct_mass;
+  if (std::strcmp(parameters.mass_solver.c_str(), "diagonal") == 0) {
+    TrilinosWrappers::SparsityPattern diag_inverse_sparsity;
+    diag_inverse_sparsity.reinit(locally_owned_dofs_fg, comm);
+
+    for (const types::global_dof_index dof : locally_owned_dofs_fg)
+      diag_inverse_sparsity.add(dof,
+                                dof);  // Add diagonal entry only
+    diag_inverse_sparsity.compress();
+
+    diag_inverse.reinit(diag_inverse_sparsity);
+    for (const types::global_dof_index dof : locally_owned_dofs_fg)
+      diag_inverse.set(dof, dof, inverse_diag_mass_squared(dof));
+    diag_inverse.compress(VectorOperation::insert);
+
+    invW = linear_operator<TrilinosWrappers::MPI::Vector,
+                           TrilinosWrappers::MPI::Vector, PayloadType>(
+        diag_inverse);
+  } else if (std::strcmp(parameters.mass_solver.c_str(), "direct") == 0) {
+    // Use direct solver from Trilinos
+    TrilinosWrappers::SolverDirect::AdditionalData data;
+    data.output_solver_details = false;
+    data.solver_type = "Amesos_Klu";
+    {
+      TimerOutput::Scope t(computing_timer, "Factorize mass matrix");
+      solver_direct_mass.initialize(mass_matrix_fg, data);
```

Comment:
If we want to wrap a direct solver (like MUMPS) with a LinearOperator, we need to add `Tvmult()`, otherwise it won't compile... I'll open a PR upstream